gk/fib.c

```diff
@@ -59,46 +64,114 @@ gk_lpm_add_ipv6_route(uint8_t *ip,
+			ref_cnt, __func__);
+	}
+
+	rte_atomic32_init(&eth_cache->ref_cnt);
 
 	return 0;
 }
 
-/*
- * Fill up the Ethernet cached header.
- * Note that the destination MAC address should be filled up by LLS.
- */
-static inline void
-fill_up_ether_cache_locked(struct ether_cache *eth_cache,
+static void
+gk_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	if (pcall_again == NULL) {
+		clear_ether_cache(eth_cache);
+		return;
+	}
+
+	/*
+	 * Deal with concurrency control by sequential lock
+	 * on the nexthop entry.
+	 */
+	write_seqlock(&eth_cache->lock);
+	ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+	eth_cache->stale = map->stale;
+	write_sequnlock(&eth_cache->lock);
+
+	*pcall_again = true;
+}
+
+/* Get a new Ethernet cached header, and fill up the header accordingly. */
+static struct ether_cache *
+get_new_ether_cache_locked(struct neighbor_hash_table *neigh,
 	struct ipaddr *addr, struct gatekeeper_if *iface)
 {
+	int i;
+	struct ether_cache *eth_cache = NULL;
+
+	for (i = 0; i < neigh->tbl_size; i++) {
+		if (rte_atomic32_read(&neigh->cache_tbl[i].ref_cnt) == 0) {
+			eth_cache = &neigh->cache_tbl[i];
+			break;
+		}
+	}
+
+	if (eth_cache == NULL)
+		return NULL;
+
+	/*
+	 * We are initializing @eth_cache, no one but us should be
+	 * reading/writing to @eth_cache, so it doesn't need a sequential lock
+	 * to protect the operations here.
+	 */
 	eth_cache->stale = true;
 	rte_memcpy(&eth_cache->ip_addr, addr, sizeof(eth_cache->ip_addr));
-	eth_cache->eth_hdr.ether_type = addr->proto;
+	eth_cache->eth_hdr.ether_type = rte_cpu_to_be_16(addr->proto);
 	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
-	eth_cache->ref_cnt = 1;
+	rte_atomic32_set(&eth_cache->ref_cnt, 1);
+
+	return eth_cache;
 }
 
 static struct ether_cache *
 neigh_get_ether_cache_locked(struct neighbor_hash_table *neigh,
-	struct ipaddr *addr, struct gatekeeper_if *iface)
+	struct ipaddr *addr, struct gatekeeper_if *iface, int lcore_id)
 {
 	int ret;
 	struct ether_cache *eth_cache = lookup_ether_cache(neigh, &addr->ip);
 	if (eth_cache != NULL) {
-		eth_cache->ref_cnt++;
+		rte_atomic32_inc(&eth_cache->ref_cnt);
 		return eth_cache;
 	}
 
-	ret = rte_hash_add_key(neigh->hash_table, &addr->ip);
-	if (ret >= 0) {
-		eth_cache = &neigh->cache_tbl[ret];
-		fill_up_ether_cache_locked(eth_cache, addr, iface);
+	eth_cache = get_new_ether_cache_locked(neigh, addr, iface);
+	if (eth_cache == NULL)
+		return NULL;
+
+	if (addr->proto == ETHER_TYPE_IPv4) {
+		ret = hold_arp(gk_arp_and_nd_req_cb,
+			eth_cache, &addr->ip.v4, lcore_id);
+	} else if (likely(addr->proto == ETHER_TYPE_IPv6)) {
+		ret = hold_nd(gk_arp_and_nd_req_cb,
+			eth_cache, &addr->ip.v6, lcore_id);
+	} else {
+		rte_panic("Unexpected condition at %s: unknown IP type %hu\n",
+			__func__, addr->proto);
```

Comment:
Given that the code already gracefully handles this failure, you can replace `rte_panic()` with `RTE_LOG(CRIT, ...)`.

Reply:
Done